### PR TITLE
Zero downtime k8s rolling delpoyments

### PR DESCRIPTION
**Pull request by NemoOudeis**

### CODE REVIEW

**Comment by NemoOudeis** [comment]
PR review -> https://jira.rakuten-it.com/jira/browse/DAF-2413

**Comment by jcayzac** [comment]
Can we have the Prism stuff in another PR?

File: content/post/graceful-k8s-delpoyments/index.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
+---
+title: "Zero-Downtime Rolling Deployments in Kubernetes"
+date: 2021-07-1500:00:00+09:00
+hero: /images/rolling-waves.jpg
```
Comment (jcayzac) [inline content/post/graceful-k8s-delpoyments/index.md:4]: Why is the hero image not in `content/post/graceful-k8s-delpoyments/` and referenced with a relative link ?

File: content/post/graceful-k8s-delpoyments/index.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+excerpt: |
+  Kubernetes enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways.
+authors:
+  - Manish Yadav
+  - Oscar Garcia
+  - Nemo
+  - Boaz Yaniv
```
Comment (jcayzac) [inline content/post/graceful-k8s-delpoyments/index.md:11]: Wow. Is there a primary author?

File: content/post/graceful-k8s-delpoyments/index.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+license: cc-by
+---
+
+*[Photo by Kammeran Gonzalez-Keola](https://www.pexels.com/@kammeran-gonzalez-keola-3137381) used under [Pexels License](https://www.pexels.com/license/)*
+
+At Rakuten we default to Kubernetes to run our services, and k8s enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways. For example, during rolling deployments we expect no request failures - a reasonable expectation, right? But when we ran functional tests during a deployment we notice requests failing - to our surprise. After researching this for a while we learned that...
+
+> K8s will send traffic to terminating pods.
+
+Even after starting the pod shutdown process, which includes sending a `TERM` signal to pod, k8s still sends new requests to the pod. Because there is no orchestration between k8s sending `TERM` signal and removing the pod from its service endpoint list. Those 2 operations can happen in any order, possibly with delay in between them. Yikes 😬.
```
Comment (jcayzac) [inline content/post/graceful-k8s-delpoyments/index.md:22]: > Even after starting the pod shutdown process, which includes sending a `TERM` signal to pod, k8s still sends new requests to the pod.

Is this correct? [Google says traffic stops as soon as the pod receives `SIGTERM`](https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace#:~:text=At%20this%20point%2C%20the%20pod%20stops%20getting%20new%20traffic.%20Containers%20running%20in%20the%20pod%20will%20not%20be%20affected.), before any `preStop` hook gets called.

File: content/authors/oscar-garcia/_index.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+title: Oscar Garcia
+bio: |
+  Cloud computing
+avatar: 	/authors/oscar-garcia/oscar_garcia.jpeg
+social:
+  - title: linkedin
+    url: https://www.linkedin.com/in/oscar-fernando-garcia-alvarado-200591181/?originalSubdomain=jp
```
Comment (jcayzac) [inline content/authors/oscar-garcia/_index.md:8]: Is this a guest post?

File: content/post/graceful-k8s-delpoyments/index.md (hunk 3 above)
Comment (jcayzac) [reply content/post/graceful-k8s-delpoyments/index.md:22]: Confirmed that Google is lying 😂

File: assets/css/prism.css [hunk 5]
```diff
@@ -1,5 +1,5 @@
-/* PrismJS 1.22.0
```
Comment (jcayzac) [inline assets/css/prism.css:1]: Can we have the Prism stuff in another PR?

**Review by jcayzac — COMMENTED** [review]
Can we have the Prism stuff in another PR?

File: content/post/graceful-k8s-delpoyments/index.md (hunk 1 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:4]: because it's also used other places so it needs an absolute link (e.g. on the home page and I think at the bottom of a post in the "more articles" section). I can move the picture inside the posts folder, but will need an absolute path regardless

File: content/post/graceful-k8s-delpoyments/index.md (hunk 2 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:11]: well, it's synthesized from several posts & discussions from discourse so I can't say who is the "main" author (maybe the only thing I can say with confidence is that I'm the editor of it^^).

File: assets/css/prism.css (hunk 5 above)
Comment (NemoOudeis) [reply assets/css/prism.css:1]: can, but I put it in here because I added new syntax highlighting necessary for this post (docker files) so we'd have to merge it first.
what's the value of splitting the PRs tho?

File: content/post/graceful-k8s-delpoyments/index.md (hunk 1 above)
Comment (jcayzac) [reply content/post/graceful-k8s-delpoyments/index.md:4]: Hugo limitation, it can't resolve files —got it!

File: assets/css/prism.css (hunk 5 above)
Comment (jcayzac) [reply assets/css/prism.css:1]: Having two different commits once we squash the PRs

File: content/post/graceful-k8s-delpoyments/index.md (hunk 2 above)
Comment (boazy) [reply content/post/graceful-k8s-delpoyments/index.md:11]: It's like Homer from the Iliad and Odyssey.
We're all Homer 🤣

File: content/post/graceful-k8s-delpoyments/index.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+## Kubernetes Pod Termination Process
+
+A pod can be evicted for multiple reasons. For example, when a node is drained, via `kubectl delete`, or when the scheduler evicts pods to allow the execution of higher-priority ones. The eviction process starts when the API server modifies the state of a pod in `etcd` to the `Terminating` state. The node’s Kubelet and the endpoints-controller are continuously watching the pod's state. Once they notice the termination state they start the eviction process:
+
+1. Kubelet performs the pod eviction.
+1. The endpoints-controller handles the endpoint removal process.
+
+Both operations are asynchronous.
+
+![SIGTERM in Kubernetes](./images/sigterm.png)
```
Comment (jcayzac) [inline content/post/graceful-k8s-delpoyments/index.md:35]: Please add `Kubelet`, `kube` and `IPtables` to your dictionary 😂

File: content/post/graceful-k8s-delpoyments/index.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+When kubelet acknowledges that a pod should be terminated, it initiates a shutdown sequence for each container in the pod. First, it runs [the container’s pre-stop hook](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution) (if it exists), then sends a `TERM` signal, and finally waits for the termination of the container. This sequential process should take less than 30 seconds (or the value in seconds specified in the [`spec.terminationGracePeriodSeconds`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#podspec-v1-core) field). If the container still running beyond this time, kubelet waits for 2 more seconds, and then kills the container forcibly by sending a `KILL` signal.
+
+![SIGKILL in Kubernetes](./images/sigkill.png)
+
+In parallel, the endpoints-controller removes the pod’s endpoint by requesting the API server. This server notifies all the kube-proxies on the worker nodes. Each kube-proxy removes the endpoint from the iptables rules in its node. Note that changes in the iptables do not affect already established connections.
```
Comment (jcayzac) [inline content/post/graceful-k8s-delpoyments/index.md:41]: I thought BPF now replaced iptables in all modern distros?

File: content/post/graceful-k8s-delpoyments/index.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+  terminationGracePeriodSeconds: 60
+  containers:
+  - name: "{{APP_NAME}}"
+    lifecycle:
+      preStop:
+        exec:
+          command: ["/bin/sh","-c","sleep 20"]
+```
+
+An important note about long lived connections: the endpoint eviction does not affect established connections. So either you need a long enough `spec.terminationGracePeriodSeconds` to terminate long-lived connections or you need ensure that connections are short lived.
```
Comment (jcayzac) [inline content/post/graceful-k8s-delpoyments/index.md:104]: Suggested:

```diff
- long lived connections
+ long lived TCP connections
```

QUIC connections rely on UDP and can be maintained across pods.

File: content/post/graceful-k8s-delpoyments/index.md (hunk 6 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:35]: not my diagram 😂
originates from oscar's post

File: assets/css/prism.css (hunk 5 above)
Comment (NemoOudeis) [reply assets/css/prism.css:1]: I split out the prism commit, at this point I don't think we have to squash the PR, because all 3 commits are stand alone & self contained (but build upon each other)

* default avatar
* prism
* article

if you prefer I can open 2 more PRs for the individual changes

File: content/post/graceful-k8s-delpoyments/index.md (hunk 7 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:41]: tbh I'm not sure and didn't find a clear answer by quick search. I'm asking one of my devops guys to confirm.

Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:41]: So yes eBPF is set to replace iptables in k8s and container networking solutions, so we have to caveat that statement there, I'll add a footnote for it. (For the established connections point - I don't know if eBPF makes a difference, we'd have to try).

The overall issue of routing state propagating through the whole cluster is _probably_ still there, I don't see how eBPF could make distributed state more consistent (just CAP wise) - but I don't know enough to give a qualified statement really.

Comment (boazy) [reply content/post/graceful-k8s-delpoyments/index.md:41]: The iptables usage we are specifically discussing is by kube-proxy. It looks like kube-proxy is not going to dump iptables but rather that modern Kubernetes cluseters can choose to dump kube-proxy itself instead. The replacement is Cilium:
https://archive.fosdem.org/2020/schedule/event/replacing_iptables_with_ebpf/attachments/slides/3622/export/events/attachments/replacing_iptables_with_ebpf/slides/3622/Cilium_FOSDEM_2020.pdf

Also:
https://cilium.io/blog/2019/08/20/cilium-16#kubeproxy-removal

The end-note should that the behavior mentioned here also applies to Cilium (or at least seems to apply to it), but there may be more caveats that you have to keep in mind. For instance, Cilium seems to have [a bug](https://github.com/cilium/cilium/issues/14844) that causes **existing connections** to also fail during pod termination, making graceful shutdown harder.
See also the [original issue](https://github.com/cilium/cilium/issues/10217).

**Review by boazy — APPROVED** [review]

**Review by boazy — CHANGES_REQUESTED** [review]
Still not approved. A sec ;)

File: content/post/graceful-k8s-delpoyments/index.md (hunk 7 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md]: I put a footnote on this part and added a closing note outlining that this is still a problem for eBPF CNI implementations. Let me know what you think of that (or if it's wrong^^)

File: assets/css/prism.css (hunk 5 above)
Comment (NemoOudeis) [reply assets/css/prism.css:1]: moved the prism stuff over into #23 so it doesn't create conflicts with additional syntax highlighting required for other posts. the diff still shows up here because this branch is on top of the other one, so the local preview renders code snippets properly.

**Review by boazy — CHANGES_REQUESTED** [review]

File: content/post/graceful-k8s-delpoyments/index.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+This is all based on Kubernetes clusters using kube-proxy & iptables. Kubernetes is moving towards [CNI implementations that use eBPF instead of iptables](https://archive.fosdem.org/2020/schedule/event/replacing_iptables_with_ebpf/) like [Cilium](https://cilium.io/) or [Calico](https://docs.projectcalico.org/about/about-calico). But the behavior described here also applies to those eBPF based solutions (or at least seems to apply to it) and there may be more caveats that you have to keep in mind. For instance, Cilium seems to have a [bug](https://github.com/cilium/cilium/issues/14844) that causes existing connections to also fail during pod termination, making graceful shutdown harder.
```
Comment (boazy) [inline content/post/graceful-k8s-delpoyments/index.md:209]: Can we use footnote syntax here?

File: content/post/graceful-k8s-delpoyments/index.md (hunk 6 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:35]: k, will do

File: content/post/graceful-k8s-delpoyments/index.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+  - Boaz Yaniv
+draft: true
+license: cc-by
+---
+
+*[Photo by Kammeran Gonzalez-Keola](https://www.pexels.com/@kammeran-gonzalez-keola-3137381) used under [Pexels License](https://www.pexels.com/license/)*
+
+At Rakuten we default to Kubernetes to run our services, and k8s enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways. For example, during rolling deployments we expect no request failures - a reasonable expectation, right? But when we ran functional tests during a deployment we notice requests failing - to our surprise. After researching this for a while we learned that...
+
+> K8s will send traffic to terminating pods.
```
[the remark this answers is not in the document]
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:20]: I'm ok with that, consistency is probably better. will do a string search for k8s....

File: content/post/graceful-k8s-delpoyments/index.md (hunk 9 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:209]: Which part exactly do you want to put in a footnote? The whole paragraph, or just the last sentence referencing the cilium bug?

File: content/post/graceful-k8s-delpoyments/index.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md]: replaced 'k8s' with 'Kubernetes' - kept it in a separate commit so we can revert if necessary

File: content/post/graceful-k8s-delpoyments/index.md (hunk 6 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md:35]: I re-created the diagram, without squiggly red lines and every word in lower case (except for all caps SIGTERM and API). lmk if you spot any other problems with the diagram.

File: content/post/graceful-k8s-delpoyments/index.md (hunk 9 above)
Comment (NemoOudeis) [reply content/post/graceful-k8s-delpoyments/index.md]: ok, on second read it actually fits better into a footnote, I moved it.

**Comment by NemoOudeis** [comment]
I did a grammerly pass over the whole thing

**Review by jcayzac — APPROVED** [review]

**Review by boazy — APPROVED** [review]
